src/ngx_http_lua_input_filters.c

```diff
@@ -0,0 +1,134 @@
+
+/*
+ * Copyright (C) by OpenResty Inc.
+ */
+
+
+#ifndef DDEBUG
+#define DDEBUG 0
+#endif
+#include "ddebug.h"
+
+
+#include "ngx_http_lua_common.h"
+
+
+ngx_int_t
+ngx_http_lua_read_bytes(ngx_buf_t *src, ngx_chain_t *buf_in, size_t *rest,
+    ssize_t bytes, ngx_log_t *log)
+{
+    if (bytes == 0) {
+        return NGX_ERROR;
+    }
+
+    if ((size_t) bytes >= *rest) {
+
+        buf_in->buf->last += *rest;
+        src->pos += *rest;
+        *rest = 0;
+
+        return NGX_OK;
+    }
+
+    /* bytes < *rest */
+
+    buf_in->buf->last += bytes;
+    src->pos += bytes;
+    *rest -= bytes;
+
+    return NGX_AGAIN;
+}
+
+
+ngx_int_t
+ngx_http_lua_read_all(ngx_buf_t *src, ngx_chain_t *buf_in, ssize_t bytes,
+    ngx_log_t *log)
+{
+    if (bytes == 0) {
+        return NGX_OK;
+    }
+
+    buf_in->buf->last += bytes;
+    src->pos += bytes;
+
+    return NGX_AGAIN;
+}
+
+
+ngx_int_t
+ngx_http_lua_read_any(ngx_buf_t *src, ngx_chain_t *buf_in, size_t *max,
+    ssize_t bytes, ngx_log_t *log)
+{
+    if (bytes == 0) {
+        return NGX_ERROR;
+    }
+
+    bytes = ((size_t) bytes >= *max) ? (ssize_t) *max : bytes;
```

Comment:
This line is not very readable. Better be

```C
if (bytes >= (ssize_t) *max) {
    bytes = (ssize_t) *max;
}
```

It's safer to convert unsigned values to signed values. Negative values can be much more troublesome when being converted into very large numbers.